File: leetCode/python/p3_LongestSubstringWithoutRepeatingCharacters/lc3.py
import time

class Solution:
    """
    Given a string s, find the length of the longest substring without repeating characters. 
    """
    def lengthOfLongestSubstring(self, s: str) -> int:
        substring = {}
        curr_substring_start = 0
        curr_len = 0         
        longest = 0
        
        for i, letter in enumerate(s): 
            if letter in substring and substring[letter] >= curr_substring_start:
                curr_substring_start = substring[letter] + 1
                curr_len = i - substring[letter] 
                substring[letter] = i
            else:
                substring[letter] = i
                curr_len += 1
                if curr_len > longest:
                    longest = curr_len

        return longest
     
    def test_function(self, s: str, sol_num: int, test_num: int):
        start_time = time.time() 
        sol = self.lengthOfLongestSubstring(s) 
        end_time = time.time()
        if(sol == sol_num):
            print('\nTEST' + str(test_num) + ': PASSED IN ' + '{:.2f}us'.format((end_time - start_time) * 1000000))
        else:
            print('\nTEST FAILED')
            
# Test 1

string = 'abcabcbb' 
s = Solution()
s.test_function(string, 3, 1)

# Test 2

string = 'bbbbb' 
s = Solution()
s.test_function(string, 1, 2)

# Test 3

string = 'pwwkew' 
s = Solution()
s.test_function(string, 3, 3)

# Test 4

string = ''
s = Solution()
s.test_function(string, 0, 4)

File: leetCode/c/p3_LongestSubstringWithoutRepeatingCharacters/lc3.c
/* Given a string s, find the length of the longest substring without repeating characters. */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

int lengthOfLongestSubstring(char * s) {

    int stringSize = strlen(s); 
    int substring[127]; // Number of characters in ASCII table (not including extended set of characters)   
    int currSubstringStart = 0;
    int currLength = 0;
    int longest = 0;

    memset(substring, -1, sizeof(substring)); // Set all array values to -1
    
    for(int i=0; i<stringSize; i++) {
        if(substring[s[i]] >= currSubstringStart) {
            currSubstringStart = substring[s[i]] + 1; 
            currLength = i - substring[s[i]];
            substring[s[i]] = i; 
        } 
        else {
            substring[s[i]] = i;
            currLength++;
            if(currLength > longest)
                longest = currLength;
        }
    }        
    
    return longest;
}

/* Test */

void testFunction(char *s, int solution, int testNumber) {
   
    int temp;
    double timeSpent;

    clock_t tic = clock();
    
    temp = lengthOfLongestSubstring(s); 
    
    clock_t toc = clock(); 
    
    timeSpent = (double) (toc - tic) * 1000000 / CLOCKS_PER_SEC;
 
    printf("\nTEST%d: ", testNumber);

    if(temp == solution) 
        printf("PASSED IN %.2fus\n", timeSpent);
    else  
        printf("FAILED\n");
}

int main(void) {
   
    char str1[] = "abcabcbb"; 
    char str2[] = "bbbbb"; 
    char str3[] = "pwwkew";
    char str4[] = "";
    
    testFunction(str1, 3, 1);
    testFunction(str2, 1, 2);
    testFunction(str3, 3, 3);
    testFunction(str4, 0, 4);
    
    return EXIT_SUCCESS;
}
